### PR TITLE
Fix span when suggesting to add an associated type bound

### LINKED ISSUE #87261: Incorrect suggestion when trying to implement `Trait<Associated = ()>` with `impl Trait + 'static`
Given the following code ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc010f95c7ce224d84a5fdd06be94dc6)): 

```rust
trait Trait {
    type Associated;
}

fn returns_opaque() -> impl Trait + 'static {
    struct Impl;

    impl Trait for Impl {
        type Associated = ();
    }

    Impl
}

fn accepts_trait<T>(_: T)
where
    T: Trait<Associated = ()>,
{
}

fn main() {
    accepts_trait(returns_opaque());
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error[E0271]: type mismatch resolving `<impl Trait as Trait>::Associated == ()`
  --> src/main.rs:22:5
   |
5  | fn returns_opaque() -> impl Trait + 'static {
   |                        -------------------- the found opaque type
...
15 | fn accepts_trait<T>(_: T)
   |    ------------- required by a bound in this
16 | where
17 |     T: Trait<Associated = ()>,
   |              --------------- required by this bound in `accepts_trait`
...
22 |     accepts_trait(returns_opaque());
   |     ^^^^^^^^^^^^^ expected `()`, found associated type
   |
   = note:    expected unit type `()`
           found associated type `<impl Trait as Trait>::Associated`
help: consider constraining the associated type `<impl Trait as Trait>::Associated` to `()`
   |
5  | fn returns_opaque() -> impl Trait + 'static<Associated = ()> {
   |                                            ^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0271`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

The syntax `impl Trait + 'static<Associated = ()>` does not actually work.

Ideally the suggested fix should look like:

```
help: consider constraining the associated type `<impl Trait as Trait>::Associated` to `()`
   |
5  | fn returns_opaque() -> impl Trait<Associated = ()> + 'static {
   |                                  ^^^^^^^^^^^^^^^^^
```

This bug appears to be present in stable, beta, and nightly, on both Rust 2015 and 2018.

### PR DESCRIPTION
**Pull request by SkiFire13**
Fixes #87261 

Note that this fix is not perfect, it ~~will still give incorrect~~ won't give suggestions in some situations:
- If the associated type is defined on a supertrait of those contained in the opaque type, it will fallback to the previous behaviour, e.g. if `AssocTy` is defined on the trait `Foo`, `Bar` has `Foo` as supertrait and the opaque type is a `impl Bar + Baz`.
- If the the associated type is defined on a generic trait and the opaque type includes two versions of that generic trait, e.g. the opaque type is `impl Foo<A> + Foo<B>`

### CODE REVIEW

**Comment by oli-obk** [comment]
Please add tests for the two issues you described

File: compiler/rustc_middle/src/ty/error.rs [hunk 1]
```diff
@@ -771,11 +771,40 @@ fn foo(&self) -> Self::T { String::new() }
...
+                    let trait_ref_def_id = bound
+                        .trait_ref()?
+                        .trait_def_id()
+                        .expect("All `hir::TraitDef`s should already be resolved");
+                    if trait_ref_def_id != assoc_trait_ref.def_id {
+                        return None;
+                    }
+                    Some(bound.span())
+                })
+                .unwrap_or_else(|| self.def_span(def_id));
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/error.rs:801]: I think we should not produce any suggestion in this case. We could still emit a help message at the def_span talking about the fact that the bound needs to be added here

Comment (SkiFire13) [reply compiler/rustc_middle/src/ty/error.rs:801]: The unfortunate thing about doing that is that we lose suggestions for cases like `impl SubTrait`. I might have had an idea to handle those cases though.

**Comment by SkiFire13** [comment]
> Please add tests for the two issues you described

Should I add tests for them even if they currently produce the wrong output? Or should I make them not suggest anything and test that?

**Comment by oli-obk** [comment]
> Or should I make them not suggest anything and test that?

Either is fine. We often test wrong output and add comments to the test stating that this is wrong and should be fixed.

**Comment by SkiFire13** [comment]
- If the code can't find the right trait to which add the bound it no longer emits a suggestion, instead it just emits the help message on the whole opaque type span
- I've added some logic to handle cases like `impl SubTrait` or `impl SubTrait + 'lt` which weren't covered before but I think are common enough it would be unfortunate to lose the correct suggestion on them.

File: src/test/ui/associated-types/issue-87261.stderr [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+   |                           --------------- required by this bound in `accepts_trait`
+...
+LL |     accepts_trait(returns_opaque_derived());
+   |     ^^^^^^^^^^^^^ expected `()`, found associated type
+   |
+   = note:    expected unit type `()`
+           found associated type `<impl DerivedTrait as Trait>::Associated`
+help: consider constraining the associated type `<impl DerivedTrait as Trait>::Associated` to `()`
+   |
+LL | fn returns_opaque_derived() -> impl DerivedTrait<Associated = ()> + 'static {
```
Comment (oli-obk) [inline src/test/ui/associated-types/issue-87261.stderr:36]: Is that legal? I thought you to add a `+ Trait<Associated = ()>`.

Comment (SkiFire13) [reply src/test/ui/associated-types/issue-87261.stderr:36]: [Yes](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f6cf553a7001efddd3e4d99d0426f2ac)

**Comment by SkiFire13** [comment]
Digging around I think I've found a similar issue with generic bounds [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0a10d192a407a8a492a5898d02db079e). The existing logic for them is a bit smarter than the one for opaque types, however it still fails when there are multiple generics. I'm tempted to include a fix for that issue in this PR and merge the logic.

File: src/test/ui/associated-types/issue-87261.stderr (hunk 2 above)
Comment (oli-obk) [reply src/test/ui/associated-types/issue-87261.stderr:36]: TIL 🤯

**Comment by oli-obk** [comment]
> I'm tempted to include a fix for that issue in this PR and merge the logic.

Whatever you prefer. We can also merge this right now and you can build a new PR on top of it

File: compiler/rustc_middle/src/ty/error.rs [hunk 3]
```diff
@@ -908,23 +921,30 @@ fn foo(&self) -> Self::T { String::new() }
         assoc_substs: &[ty::GenericArg<'tcx>],
         ty: Ty<'tcx>,
         msg: &str,
+        is_bound_surely_present: bool,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/error.rs:924]: Please document this argument's effect. It is clear in the context of this PR, but probably not for anyone just stumbling across it.

**Comment by oli-obk** [comment]
@bors r+

Thanks!

**Pull request merged**